de.metas.business/src/main/java-legacy/org/compiere/process/BOMValidate.java

```diff
@@ -93,194 +82,195 @@ else if (I_PP_Product_BOM.Table_Name.equals(tableName))
-			else
-				pstmt.setInt(1, p_M_Product_Category_ID);
-			rs = pstmt.executeQuery ();
-			while (rs.next ())
-			{
-				String info = validateProduct(new MProduct(getCtx(), rs.getInt(1), get_TrxName()));
-				addLog(0, null, null, info);
-				counter++;
-			}
+			queryBuilder.addEqualsFilter(I_M_Product.COLUMNNAME_M_Product_Category_ID, p_M_Product_Category_ID);
 		}
-		catch (Exception e)
+		else
 		{
-			log.error(sql, e);
+			queryBuilder.addOnlyContextClient();
 		}
-		finally
+
+		if (!p_IsReValidate)
 		{
-		  DB.close(rs, pstmt);
-		  rs = null; pstmt = null;
+			queryBuilder.addNotEqualsFilter(I_M_Product.COLUMNNAME_IsVerified, true);
 		}
-		return "#" + counter;
-	}	//	doIt
 
-	
-	
+		queryBuilder.orderBy()
+				.addColumn(I_M_Product.COLUMNNAME_Name)
+				.endOrderBy();
+
+		final AtomicInteger counter = new AtomicInteger(0);
+
+		queryBuilder.create()
+				.stream()
+				.forEach(product -> {
+					String info = validateProduct(product);
+					addLog(0, null, null, info);
+					counter.incrementAndGet();
+				});
+
+
+		return "#" + counter.get();
+	}	// doIt
+
+
 	/**
-	 * 	Validate Product
-	 *	@param product product
-	 *	@return Info
+	 * Validate Product
+	 *
+	 * @param product product
+	 * @return Info
 	 */
-	private String validateProduct (MProduct product)
+	private String validateProduct(@NonNull final I_M_Product product)
 	{
 		if (!product.isBOM())
+		{
 			return product.getName() + " @NotValid@ @M_BOM_ID@";
+		}
 		m_product = product;
-	
-		//	Check Old Product BOM Structure
-		log.info(m_product.getName());
-		m_products = new ArrayList<MProduct>();
-		if (!validateOldProduct (m_product))
+
+		// Check Old Product BOM Structure
+		m_products = new ArrayList<>();
+		if (!validateOldProduct(m_product))
 		{
-			m_product.setIsVerified(false);
-			m_product.save();
+			saveVerifiedProduct(m_product, false);
 			return m_product.getName() + " @NotValid@";
 		}
-		
-		//	New Structure
+
+		// New Structure
 		MBOM[] boms = MBOM.getOfProduct(getCtx(), p_M_Product_ID, get_TrxName(), null);
 		for (int i = 0; i < boms.length; i++)
 		{
-			m_products = new ArrayList<MProduct>();
+			m_products = new ArrayList<>();
 			if (!validateBOM(boms[i]))
 			{
-				m_product.setIsVerified(false);
-				m_product.save();
+				saveVerifiedProduct(m_product, false);
 				return m_product.getName() + " " + boms[i].getName() + " @NotValid@";
 			}
 		}
-		
-		//	OK
-		m_product.setIsVerified(true);
-		m_product.save();
-		return m_product.getName() + " @IsValid@";
-	}	//	validateProduct
 
+		// OK
+		saveVerifiedProduct(m_product, true);
+		return m_product.getName() + " @IsValid@";
+	}	// validateProduct
 
 	/**
-	 * 	Validate Old BOM Product structure
-	 *	@param product product
-	 *	@return true if valid
+	 * Validate Old BOM Product structure
+	 *
+	 * @param product product
+	 * @return true if valid
 	 */
-	private boolean validateOldProduct (MProduct product)
+	private boolean validateOldProduct(@NonNull final I_M_Product product)
```

Comment:
imho we shall drop this!
@metas-ts  wdyt?

Reply:
we no longer use the old BOM format... actually we metasfresh never used that...